### PR TITLE
Enable ruff linter, fix all lint warnings, disable the rest

### PR DESCRIPTION
**Pull request by sritchie**
This PR:

- enables Ruff's linting mode (this picks up the config options we already had in pyproject.toml, what a win)
- adds github-formatted Ruff outputs for any linter bugs that make it past the pre-commit hooks
- fixes all lint errors, and ignores 2 error types (comment added in pyproject.toml)

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -24,6 +24,5 @@ repos:
 - repo: https://github.com/astral-sh/ruff-pre-commit
   rev: v0.1.11
   hooks:
-    # TODO enable this when we're prepared to clean up ruff's linter warnings.
-    # - id: ruff
+    - id: ruff
```
Comment (sritchie) [inline .pre-commit-config.yaml:27]: This does in fact pick up config settings from `pyproject.toml`.

File: pyproject.toml [hunk 2]
```diff
@@ -105,7 +105,11 @@ indent-width = 4
 [tool.ruff.lint]
 extend-select = ["I"]
 select = ["E4", "E7", "E9", "F"]
-ignore = []
+# F811 silences errors from GenJAX's type-based dispatch. Remove this if we can
+# find some way to teach ruff about these overloads.
+#
+# F403 disables errors from `*` imports, which we currently use heavily.
```
Comment (sritchie) [inline pyproject.toml:111]: I wasn't sure how you wanted to handle these. This causes:

```
src/genjax/__init__.py:22:1: F403 `from .console import *` used; unable to detect undefined names
src/genjax/__init__.py:23:1: F403 `from .core import *` used; unable to detect undefined names
src/genjax/__init__.py:24:1: F403 `from .experimental import *` used; unable to detect undefined names
src/genjax/__init__.py:25:1: F403 `from .extras import *` used; unable to detect undefined names
src/genjax/__init__.py:26:1: F403 `from .generative_functions import *` used; unable to detect undefined names
src/genjax/__init__.py:27:1: F403 `from .inference import *` used; unable to detect undefined names
src/genjax/__init__.py:28:1: F403 `from .information import *` used; unable to detect undefined names
src/genjax/_src/extras/blackjax/__init__.py:15:1: F403 `from .blackjax import *` used; unable to detect undefined names
src/genjax/_src/extras/tinygp/__init__.py:15:1: F403 `from .tinygp import *` used; unable to detect undefined names
src/genjax/core/__init__.py:15:1: F403 `from .datatypes import *` used; unable to detect undefined names
src/genjax/core/__init__.py:16:1: F403 `from .interpreters import *` used; unable to detect undefined names
src/genjax/core/__init__.py:17:1: F403 `from .pytree import *` used; unable to detect undefined names
src/genjax/generative_functions/__init__.py:15:1: F403 `from genjax._src.generative_functions.drop_arguments import *` used; unable to detect undefined names
src/genjax/generative_functions/__init__.py:17:1: F403 `from .combinators import *` used; unable to detect undefined names
src/genjax/generative_functions/__init__.py:18:1: F403 `from .distributions import *` used; unable to detect undefined names
src/genjax/generative_functions/__init__.py:19:1: F403 `from .interpreted import *` used; unable to detect undefined names
src/genjax/generative_functions/__init__.py:20:1: F403 `from .static import *` used; unable to detect undefined names
Found 17 errors.
```

File: pyproject.toml [hunk 3]
```diff
@@ -105,7 +105,11 @@ indent-width = 4
 [tool.ruff.lint]
 extend-select = ["I"]
 select = ["E4", "E7", "E9", "F"]
-ignore = []
+# F811 silences errors from GenJAX's type-based dispatch. Remove this if we can
```
Comment (sritchie) [inline pyproject.toml:108]: This causes:

```
(genjax) [sritchie@wintermute ~/code/python/genjax (sritchie/enable_linting)]$ ruff .
src/genjax/_src/core/datatypes/generative.py:218:9: F811 Redefinition of unused `new` from line 210
src/genjax/_src/core/datatypes/generative.py:327:9: F811 Redefinition of unused `filter` from line 323
src/genjax/_src/core/datatypes/generative.py:365:9: F811 Redefinition of unused `filter` from line 358
src/genjax/_src/core/datatypes/generative.py:372:9: F811 Redefinition of unused `filter` from line 365
src/genjax/_src/core/datatypes/generative.py:446:9: F811 Redefinition of unused `__getitem__` from line 434
src/genjax/_src/core/datatypes/generative.py:573:9: F811 Redefinition of unused `project` from line 566
src/genjax/_src/core/datatypes/generative.py:580:9: F811 Redefinition of unused `project` from line 573
src/genjax/_src/core/datatypes/generative.py:618:9: F811 Redefinition of unused `update` from line 608
src/genjax/_src/core/datatypes/generative.py:889:9: F811 Redefinition of unused `__eq__` from line 882
src/genjax/_src/core/datatypes/generative.py:904:9: F811 Redefinition of unused `__getitem__` from line 896
src/genjax/_src/core/datatypes/generative.py:1108:9: F811 Redefinition of unused `importance` from line 1077
src/genjax/_src/core/datatypes/generative.py:1136:9: F811 Redefinition of unused `update` from line 1126
src/genjax/_src/core/datatypes/generative.py:1288:9: F811 Redefinition of unused `new` from line 1274
src/genjax/_src/core/datatypes/generative.py:1297:9: F811 Redefinition of unused `new` from line 1288
src/genjax/_src/core/datatypes/generative.py:1344:9: F811 Redefinition of unused `get_submap` from line 1339
src/genjax/_src/core/datatypes/generative.py:1349:9: F811 Redefinition of unused `get_submap` from line 1344
src/genjax/_src/core/datatypes/generative.py:1399:9: F811 Redefinition of unused `merge` from line 1384
src/genjax/_src/core/datatypes/generative.py:1403:9: F811 Redefinition of unused `merge` from line 1399
src/genjax/_src/core/datatypes/generative.py:1407:9: F811 Redefinition of unused `merge` from line 1403
src/genjax/_src/core/datatypes/generative.py:1530:5: F811 Redefinition of unused `choice_map` from line 1525
src/genjax/_src/core/datatypes/generative.py:1535:5: F811 Redefinition of unused `choice_map` from line 1530
src/genjax/_src/core/datatypes/generative.py:1540:5: F811 Redefinition of unused `choice_map` from line 1535
src/genjax/_src/core/datatypes/generative.py:1556:5: F811 Redefinition of unused `select` from line 1551
src/genjax/_src/core/datatypes/generative.py:1561:5: F811 Redefinition of unused `select` from line 1556
src/genjax/_src/generative_functions/combinators/switch/switch_datatypes.py:33:64: F811 Redefinition of unused `Sequence` from line 16
src/genjax/_src/generative_functions/combinators/vector/map_combinator.py:99:9: F811 Redefinition of unused `maybe_restore_arguments_project` from line 91
src/genjax/_src/generative_functions/combinators/vector/map_combinator.py:123:9: F811 Redefinition of unused `project` from line 110
src/genjax/_src/generative_functions/combinators/vector/map_combinator.py:258:9: F811 Redefinition of unused `importance` from line 233
src/genjax/_src/generative_functions/combinators/vector/map_combinator.py:283:9: F811 Redefinition of unused `importance` from line 258
src/genjax/_src/generative_functions/combinators/vector/map_combinator.py:294:9: F811 Redefinition of unused `importance` from line 283
src/genjax/_src/generative_functions/combinators/vector/map_combinator.py:316:9: F811 Redefinition of unused `maybe_restore_arguments_kernel_update` from line 304
src/genjax/_src/generative_functions/combinators/vector/map_combinator.py:368:9: F811 Redefinition of unused `update` from line 327
src/genjax/_src/generative_functions/combinators/vector/map_combinator.py:398:9: F811 Redefinition of unused `update` from line 368
src/genjax/_src/generative_functions/combinators/vector/unfold_combinator.py:121:9: F811 Redefinition of unused `project` from line 107
src/genjax/_src/generative_functions/combinators/vector/unfold_combinator.py:345:9: F811 Redefinition of unused `importance` from line 283
src/genjax/_src/generative_functions/combinators/vector/unfold_combinator.py:410:9: F811 Redefinition of unused `importance` from line 345
src/genjax/_src/generative_functions/combinators/vector/unfold_combinator.py:423:9: F811 Redefinition of unused `importance` from line 410
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:78:9: F811 Redefinition of unused `new` from line 72
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:84:9: F811 Redefinition of unused `new` from line 78
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:94:9: F811 Redefinition of unused `has_addr` from line 90
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:130:9: F811 Redefinition of unused `new` from line 124
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:146:9: F811 Redefinition of unused `new` from line 130
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:152:9: F811 Redefinition of unused `new` from line 146
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:167:9: F811 Redefinition of unused `filter` from line 160
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:181:9: F811 Redefinition of unused `has_submap` from line 177
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:200:9: F811 Redefinition of unused `get_submap` from line 186
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:207:9: F811 Redefinition of unused `get_submap` from line 200
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:215:9: F811 Redefinition of unused `get_submap` from line 207
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:266:9: F811 Redefinition of unused `new` from line 258
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:275:9: F811 Redefinition of unused `new` from line 266
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:300:9: F811 Redefinition of unused `filter` from line 287
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:322:9: F811 Redefinition of unused `has_submap` from line 315
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:335:9: F811 Redefinition of unused `get_submap` from line 330
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:345:9: F811 Redefinition of unused `get_submap` from line 335
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:358:9: F811 Redefinition of unused `get_submap` from line 345
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:370:9: F811 Redefinition of unused `merge` from line 365
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:385:9: F811 Redefinition of unused `merge` from line 370
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:404:5: F811 Redefinition of unused `select` from line 30
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:409:5: F811 Redefinition of unused `choice_map` from line 28
src/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:418:5: F811 Redefinition of unused `choice_map` from line 409
src/genjax/_src/generative_functions/distributions/distribution.py:152:9: F811 Redefinition of unused `importance` from line 142
src/genjax/_src/generative_functions/distributions/distribution.py:188:9: F811 Redefinition of unused `update` from line 164
src/genjax/_src/generative_functions/interpreted/__init__.py:367:9: F811 Redefinition of unused `update` from line 341
src/genjax/_src/generative_functions/static/static_transforms.py:57:70: F811 Redefinition of unused `Trie` from line 29
src/genjax/_src/generative_functions/supports_callees.py:71:9: F811 Redefinition of unused `__call__` from line 67
src/genjax/_src/gensp/core.py:89:9: F811 Redefinition of unused `importance` from line 81
src/genjax/_src/gensp/core.py:103:5: F811 Redefinition of unused `target` from line 95
src/genjax/_src/gensp/grasp.py:94:1: F811 Redefinition of unused `flip_enum` from line 25
src/genjax/_src/gensp/grasp.py:99:1: F811 Redefinition of unused `flip_mvd` from line 26
src/genjax/_src/gensp/grasp.py:115:1: F811 Redefinition of unused `normal_reinforce` from line 29
src/genjax/_src/gensp/grasp.py:120:1: F811 Redefinition of unused `normal_reparam` from line 30
src/genjax/_src/gensp/grasp.py:125:1: F811 Redefinition of unused `mv_normal_diag_reparam` from line 28
src/genjax/_src/gensp/grasp.py:140:1: F811 Redefinition of unused `geometric_reinforce` from line 27
src/genjax/_src/gensp/grasp.py:145:1: F811 Redefinition of unused `uniform` from line 59
src/genjax/_src/gensp/grasp.py:171:5: F811 Redefinition of unused `baseline` from line 24
src/genjax/_src/gensp/grasp.py:380:5: F811 Redefinition of unused `sir` from line 375
src/genjax/_src/gensp/grasp.py:389:5: F811 Redefinition of unused `sir` from line 380
src/genjax/_src/gensp/grasp.py:497:5: F811 Redefinition of unused `marginal` from line 488
src/genjax/_src/gensp/grasp.py:547:5: F811 Redefinition of unused `q_wake` from line 528
src/genjax/_src/gensp/grasp.py:582:5: F811 Redefinition of unused `p_wake` from line 562
src/genjax/_src/gensp/grasp.py:598:5: F811 Redefinition of unused `elbo` from line 510
src/genjax/_src/gensp/grasp.py:631:5: F811 Redefinition of unused `iwae_elbo` from line 612
src/genjax/_src/gensp/inference/importance_sampling.py:156:5: F811 Redefinition of unused `importance_sampler` from line 151
src/genjax/_src/inference/importance_sampling.py:254:5: F811 Redefinition of unused `importance_sampling` from line 246
src/genjax/_src/inference/importance_sampling.py:271:5: F811 Redefinition of unused `sampling_importance_resampling` from line 263
src/genjax/_src/inference/smc/init.py:114:5: F811 Redefinition of unused `smc_initialize` from line 61
src/genjax/_src/inference/smc/resample.py:81:9: F811 Redefinition of unused `apply` from line 61
src/genjax/_src/inference/smc/update.py:109:5: F811 Redefinition of unused `smc_update` from line 67
src/genjax/_src/inference/translator.py:204:5: F811 Redefinition of unused `extending_trace_translator` from line 183
src/genjax/core/runtime_debugger.py:59:5: F811 Redefinition of unused `record` from line 51
tests/generative_functions/test_switch_combinator.py:74:9: F811 Redefinition of unused `test_switch_simulate_in_gen_fn` from line 23
tests/inference/test_translators.py:90:7: F811 Redefinition of unused `TestExtendingTraceTranslator` from line 25
Found 92 errors.
```

File: src/genjax/__init__.py [hunk 4]
```diff
@@ -55,6 +52,6 @@
 ####################################################
 
 try:
-    del genjax._src
+    del genjax._src  # noqa: F405
```
Comment (sritchie) [inline src/genjax/__init__.py]: not sure if there's a better way to deal with this so I pulled the ripcord.

File: src/genjax/_src/console.py [hunk 5]
```diff
@@ -115,7 +115,7 @@ def console(
             ],
             **pretty_kwargs,
         }
-    except:
+    except Exception:
```
Comment (sritchie) [inline src/genjax/_src/console.py:118]: Do we know a more fine-grained type we except to see?

File: tests/generative_functions/test_interpreted_language.py [hunk 6]
```diff
@@ -362,8 +362,8 @@ def simple_normal():
         key, sub_key = jax.random.split(key)
         (updated, w, _, discard) = simple_normal.update(sub_key, tr, new, ())
         updated_chm = updated.get_choices()
-        y1 = updated_chm[("y1",)]
-        y2 = updated_chm[("y2",)]
+        _y1 = updated_chm[("y1",)]
```
Comment (sritchie) [inline tests/generative_functions/test_interpreted_language.py:365]: Some of these we may want to delete; these are unused accesses, different than unused traces. Let me know and I'll kill some.

File: tests/generative_functions/test_interpreted_language.py [hunk 7]
```diff
@@ -733,7 +733,7 @@ def higher_higher_model():
             y = higher_model.inline()
             return y
 
-        key = jax.random.PRNGKey(314159)
+        _key = jax.random.PRNGKey(314159)
```
Comment (sritchie) [inline tests/generative_functions/test_interpreted_language.py:736]: This was one of a few unused keys. Could this be a bigger bug?

File: tests/generative_functions/test_static_language.py [hunk 8]
```diff
@@ -589,7 +589,7 @@ def outlier(prob):
 
         key = jax.random.PRNGKey(314159)
         proposal = make_gen_fn()
-        tr = proposal.simulate(key, (0.3,))
+        _tr = proposal.simulate(key, (0.3,))
```
Comment (sritchie) [inline tests/generative_functions/test_static_language.py:592]: Is this test checking that simulate runs at all? Can we remove the assignment and just run it for side effects?

File: tests/generative_functions/test_static_language.py [hunk 9]
```diff
@@ -705,7 +705,7 @@ def higher_higher_model():
             y = higher_model.inline()
             return y
 
-        key = jax.random.PRNGKey(314159)
+        _key = jax.random.PRNGKey(314159)
```
Comment (sritchie) [inline tests/generative_functions/test_static_language.py:708]: unused!

File: tests/generative_functions/test_switch_combinator.py [hunk 10]
```diff
@@ -34,18 +34,18 @@ def model():
...
-        tr = model.simulate(sub_key, ())
+        _tr = model.simulate(sub_key, ())
         assert True
 
     def test_switch_simulate(self):
         @genjax.Static
         def simple_normal():
-            y1 = genjax.trace("y1", genjax.normal)(0.0, 1.0)
-            y2 = genjax.trace("y2", genjax.normal)(0.0, 1.0)
+            _y1 = genjax.trace("y1", genjax.normal)(0.0, 1.0)
```
Comment (sritchie) [inline tests/generative_functions/test_switch_combinator.py:43]: I like keeping these names for documentation with the underscore to show they're not meant to be used.... wdyt?

File: tests/inference/smc/test_basic_smc.py [hunk 11]
```diff
@@ -104,15 +104,15 @@ def _inner(carry, xs):
             return stacked
 
         key = jax.random.PRNGKey(314159)
-        smc_state = jax.jit(extend_smc_no_resampling)(key, obs, 0.0)
+        _smc_state = jax.jit(extend_smc_no_resampling)(key, obs, 0.0)
```
Comment (sritchie) [inline tests/inference/smc/test_basic_smc.py:107]: how about this one? meant to be used?

File: tests/inference/test_translators.py [hunk 12]
```diff
@@ -26,12 +28,13 @@ def test_extending_trace_translator_vs_manual_update(self):
...
         def model(z):
             z = genjax.normal(z, 1.0) @ "z"
-            x = genjax.normal(z, 1.0) @ "x"
+            _x = genjax.normal(z, 1.0) @ "x"
             return z
 
         @genjax.Static
         @typecheck
         def proposal(obs_chm: ChoiceMap, prev_particle: ChoiceMap, *args):
+            t = None
```
Comment (sritchie) [inline tests/inference/test_translators.py]: THIS one seemed like a real bug — we use a `t` that is never supplied. I added this to fix the linter but we need to address this one before merging. What should this be?

**Review by sritchie — COMMENTED** [review]
Some notes about potential bugs in the lint errors.

File: src/genjax/__init__.py (hunk 4 above)
Comment (femtomc) [reply src/genjax/__init__.py]: Let's just rip that chunk out -- it's a dumb thing I put in, and doesn't even work as far as I can tell.

File: src/genjax/_src/console.py (hunk 5 above)
Comment (femtomc) [reply src/genjax/_src/console.py:118]: No, unfortunately: the overload here activates `rich.traceback` -- which has pretty printing for all exceptions. We're not being specific.

File: pyproject.toml (hunk 2 above)
Comment (femtomc) [reply pyproject.toml:111]: All the `from foo import *` occur in the public API -- so it's a small set of files. But yes, we should fix -- although it's going to mess up the tests.

File: tests/generative_functions/test_interpreted_language.py (hunk 7 above)
Comment (femtomc) [reply tests/generative_functions/test_interpreted_language.py:736]: No -- I changed the signature of `assess`, which previously required a key, and didn't prune the keys.

File: tests/generative_functions/test_static_language.py (hunk 8 above)
Comment (femtomc) [reply tests/generative_functions/test_static_language.py:592]: Yes

File: tests/inference/test_translators.py (hunk 12 above)
Comment (femtomc) [reply tests/inference/test_translators.py]: `t` is closed over from `get_translator`

Comment (femtomc) [reply tests/inference/test_translators.py]: Oops, I see -- the `proposal` here is never used

Comment (sritchie) [reply tests/inference/test_translators.py]: It's used on line 82, looks like

**Pull request merged**